python/pyspark/sql/pandas/serializers.py

```diff
@@ -51,7 +55,20 @@ def load_stream(self, stream):
         """
         # load the batches
         for batch in self.serializer.load_stream(stream):
-            yield batch
+            if self.split_batches:
+                import pyarrow as pa
+                # When spark.sql.execution.arrow.pyspark.selfDestruct.enabled, ensure
+                # each column in each record batch is contained in its own allocation.
+                # Otherwise, selfDestruct does nothing; it frees each column as its
+                # converted, but each column will actually be a list of slices of record
+                # batches, and so no memory is actually freed until all columns are
+                # converted.
+                split_batch = pa.RecordBatch.from_arrays([
+                    pa.concat_arrays([array]) for array in batch
```

Comment:
I'm not crazy about using `pa.concat_arrays()` to force a copy. I don't think it's safe to assume it will always make a copy, it might end up producing a chunked array in the future with zero copy, as is done in `pa.concat_tables()`. Is there another, more direct way we can make copies?

Reply:
I'll look around. I couldn't find another obvious one but I agree that this is not ideal.

Reply:
I don't see an explicit way to copy. The take/filter kernels might suffice but would require constructing auxiliary arrays; we could also try copying buffers directly. Would it be better perhaps to defer this and work on a explicit copy in Arrow, or a RecordBatchReader that doesn't require copies?

Reply:
I don't think we need to defer since what you have works, but if you could look into those 2 improvements on the Arrow side that would be great. Then we could come back and update this later.

Reply:
I filed https://issues.apache.org/jira/browse/ARROW-10670.